arbnode/simple_redis_lock.go

```diff
@@ -0,0 +1,204 @@
+package arbnode
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/go-redis/redis/v8"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+)
+
+type SimpleRedisLock struct {
+	stopwaiter.StopWaiter
+	client      redis.UniversalClient
+	config      *SimpleRedisLockConfig
+	lockedUntil int64
+	mutex       sync.Mutex
+	stopping    bool
+	readyToLock func() bool
+	myId        string
+}
+
+type SimpleRedisLockConfig struct {
+	RedisUrl        string        `koanf:"redis-url"`
+	MyId            string        `koanf:"my-id"`
+	LockoutDuration time.Duration `koanf:"lockout-duration"`
+	RefreshDuration time.Duration `koanf:"refresh-duration"`
+	Key             string        `koanf:"key"`
+	BackgroundLock  bool          `koanf:"background-lock"`
+}
+
+func RedisLockConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".redis-url", DefaultRedisLockConfig.RedisUrl, "URL of redis server for coordination")
+	f.String(prefix+".my-id", DefaultRedisLockConfig.RedisUrl, "this node's id prefix when acquiring the lock (optional)")
+	f.Duration(prefix+".lockout-duration", DefaultRedisLockConfig.LockoutDuration, "how long lock is held")
+	f.Duration(prefix+".refresh-duration", DefaultRedisLockConfig.RefreshDuration, "how long between consecutive calls to redis")
+	f.String(prefix+".key", prefix+"simple-lock", "key for lock")
+	f.Bool(prefix+".background-lock", DefaultRedisLockConfig.BackgroundLock, "should node always try grabing lock in background")
+}
+
+func NewSimpleRedisLock(config *SimpleRedisLockConfig, readyToLock func() bool) (*SimpleRedisLock, error) {
+	var client redis.UniversalClient
+	if config.RedisUrl == "" {
+		client = nil
+	} else {
+		redisOptions, err := redis.ParseURL(config.RedisUrl)
+		if err != nil {
+			return nil, err
+		}
+		client = redis.NewClient(redisOptions)
+	}
+	return &SimpleRedisLock{
+		myId:        config.MyId + "-" + strconv.FormatInt(rand.Int63(), 16), // unique even if config is not
```

Comment:
I think `rand.Int63()` is deterministic by default. Likely other things will be calling rand in a non-deterministic way, but to avoid conflicts, we should either use a custom seeded instance or do something like get 8 cryptographically secure random bytes and hex encode them.